File: DailyBrain/python/meetings.py
# python/meetings.py
"""
MeetingsTask: Manages meetings and preparation checklists.
Integrates with Google Calendar / Outlook (stub implementation).
"""
from dataclasses import dataclass
from datetime import datetime, time, timedelta
from typing import List, Optional

@dataclass
class MeetingItem:
    id: str
    title: str
    start_time: datetime
    end_time: datetime
    location: str  # "zoom", "teams", "in_person", "phone"
    attendees: List[str]
    agenda: Optional[str] = None
    prep_checklist: Optional[List[str]] = None
    notes: Optional[str] = None

@dataclass
class MeetingsContext:
    meetings: List[MeetingItem]
    current_time: datetime
    prep_warning_minutes: int = 30

@dataclass
class MeetingsReviewResult:
    next_meeting: Optional[MeetingItem]
    today_meetings: List[MeetingItem]
    needs_prep: List[MeetingItem]
    minutes_until_next: int
    note: str

class MeetingsTask:
    def run(self, ctx: MeetingsContext) -> MeetingsReviewResult:
        """
        Reviews meetings and identifies preparation needs.
        """
        today = ctx.current_time.date()
        today_meetings = [
            m for m in ctx.meetings 
            if m.start_time.date() == today and m.start_time > ctx.current_time
        ]
        today_meetings.sort(key=lambda x: x.start_time)
        
        next_meeting = today_meetings[0] if today_meetings else None
        
        # Find meetings needing prep (within warning window)
        prep_threshold = ctx.current_time + timedelta(minutes=ctx.prep_warning_minutes)
        needs_prep = [
            m for m in today_meetings
            if m.start_time <= prep_threshold and m.prep_checklist
        ]
        
        # Calculate minutes until next meeting
        if next_meeting:
            delta = next_meeting.start_time - ctx.current_time
            minutes_until_next = int(delta.total_seconds() / 60)
        else:
            minutes_until_next = -1
        
        # Generate summary
        notes = []
        if needs_prep:
            notes.append(f"📋 {len(needs_prep)} meeting(s) need prep NOW!")
        if next_meeting:
            notes.append(f"⏰ Next: '{next_meeting.title}' in {minutes_until_next} min.")
        notes.append(f"📅 {len(today_meetings)} meeting(s) remaining today.")
        
        note = " ".join(notes) if notes else "🎉 No more meetings today!"
        
        return MeetingsReviewResult(
            next_meeting=next_meeting,
            today_meetings=today_meetings,
            needs_prep=needs_prep,
            minutes_until_next=minutes_until_next,
            note=note
        )
    
    def format_meeting(self, meeting: MeetingItem) -> str:
        """
        Format a single meeting with all details.
        """
        location_emoji = {
            "zoom": "📹", "teams": "👥", "in_person": "🏢", "phone": "📞"
        }.get(meeting.location, "📍")
        
        lines = [
            f"=== {meeting.title} ===",
            f"  {location_emoji} {meeting.location.upper()}",
            f"  🕐 {meeting.start_time.strftime('%H:%M')} - {meeting.end_time.strftime('%H:%M')}",
            f"  👤 Attendees: {', '.join(meeting.attendees)}"
        ]
        
        if meeting.agenda:
            lines.append(f"  📝 Agenda: {meeting.agenda}")
        
        if meeting.prep_checklist:
            lines.append("  ✅ Prep Checklist:")
            for item in meeting.prep_checklist:
                lines.append(f"     - {item}")
        
        return "\n".join(lines)
    
    def format_day_schedule(self, meetings: List[MeetingItem]) -> str:
        """
        Format all meetings for the day.
        """
        if not meetings:
            return "=== Today's Schedule ===\nNo meetings scheduled."
        
        lines = ["=== Today's Schedule ==="]
        for m in sorted(meetings, key=lambda x: x.start_time):
            duration = (m.end_time - m.start_time).total_seconds() / 60
            lines.append(
                f"  {m.start_time.strftime('%H:%M')} - {m.title} ({int(duration)} min)"
            )
        
        return "\n".join(lines)

File: DailyBrain/rust/src/meetings.rs
// rust/src/meetings.rs
use chrono::{DateTime, Duration, Local};

#[derive(Clone, Debug)]
pub struct MeetingItem {
    pub id: String,
    pub title: String,
    pub start_time: DateTime<Local>,
    pub end_time: DateTime<Local>,
    pub location: String, // "zoom", "teams", "in_person", "phone"
    pub attendees: Vec<String>,
    pub agenda: Option<String>,
    pub prep_checklist: Option<Vec<String>>,
}

pub struct MeetingsContext {
    pub meetings: Vec<MeetingItem>,
    pub current_time: DateTime<Local>,
    pub prep_warning_minutes: i64,
}

pub struct MeetingsReviewResult {
    pub next_meeting: Option<MeetingItem>,
    pub today_meetings: Vec<MeetingItem>,
    pub needs_prep: Vec<MeetingItem>,
    pub minutes_until_next: i64,
    pub note: String,
}

pub struct MeetingsTask;

impl MeetingsTask {
    pub fn run(&self, ctx: &MeetingsContext) -> MeetingsReviewResult {
        let today = ctx.current_time.date_naive();
        let mut today_meetings: Vec<_> = ctx
            .meetings
            .iter()
            .filter(|m| {
                m.start_time.date_naive() == today && m.start_time > ctx.current_time
            })
            .cloned()
            .collect();

        today_meetings.sort_by_key(|m| m.start_time);

        let next_meeting = today_meetings.first().cloned();

        let prep_threshold = ctx.current_time + Duration::minutes(ctx.prep_warning_minutes);
        let needs_prep: Vec<_> = today_meetings
            .iter()
            .filter(|m| m.start_time <= prep_threshold && m.prep_checklist.is_some())
            .cloned()
            .collect();

        let minutes_until_next = if let Some(ref next) = next_meeting {
            (next.start_time - ctx.current_time).num_minutes()
        } else {
            -1
        };

        let mut notes = Vec::new();
        if !needs_prep.is_empty() {
            notes.push(format!("📋 {} meeting(s) need prep NOW!", needs_prep.len()));
        }
        if let Some(ref next) = next_meeting {
            notes.push(format!(
                "⏰ Next: '{}' in {} min.",
                next.title, minutes_until_next
            ));
        }
        notes.push(format!(
            "📅 {} meeting(s) remaining today.",
            today_meetings.len()
        ));

        let note = if notes.is_empty() {
            "🎉 No more meetings today!".to_string()
        } else {
            notes.join(" ")
        };

        MeetingsReviewResult {
            next_meeting,
            today_meetings,
            needs_prep,
            minutes_until_next,
            note,
        }
    }

    pub fn format_meeting(&self, meeting: &MeetingItem) -> String {
        let location_emoji = match meeting.location.as_str() {
            "zoom" => "📹",
            "teams" => "👥",
            "in_person" => "🏢",
            "phone" => "📞",
            _ => "📍",
        };

        let mut lines = vec![
            format!("=== {} ===", meeting.title),
            format!("  {} {}", location_emoji, meeting.location.to_uppercase()),
            format!(
                "  🕐 {} - {}",
                meeting.start_time.format("%H:%M"),
                meeting.end_time.format("%H:%M")
            ),
            format!("  👤 Attendees: {}", meeting.attendees.join(", ")),
        ];

        if let Some(ref agenda) = meeting.agenda {
            lines.push(format!("  📝 Agenda: {}", agenda));
        }

        if let Some(ref checklist) = meeting.prep_checklist {
            lines.push("  ✅ Prep Checklist:".to_string());
            for item in checklist {
                lines.push(format!("     - {}", item));
            }
        }

        lines.join("\n")
    }
}
